OIDC Middleware library integrated into express sample app

vijetmahabaleshwar-okta opened the pull request:
To test the integration, you will need the following -

1. Updated angular js front end (https://github.com/okta/samples-js-angular-1/pull/24)
2. .samples.config.json should point to a real okta server / start mock-server to playback tapes
3. client id/secret in .samples.config.json should refer to "Web App" in your okta org / left unchanged if using mock-server

Currently the tests won't work as the behavior of the express backend has changed after the integration.
Test changes would be added in a separate PR
The sample app now works with mock server as well. New tapes are recorded, which is added to this PR.

rchild-okta commented on lib/index.js:
```diff
@@ -43,15 +43,29 @@ app.use(session({
 // authorization code scenarios
 app.get('/', handlers.scenarios);
 app.get('/authorization-code/login-redirect', handlers.loginRedirect);
-app.get('/authorization-code/login-custom', handlers.loginCustom);
```
In this case we still don't have an option for login-custom right?

vijetmahabaleshwar-okta replied:
Yes, we don't have the custom login option yet, but I think we need to discuss how do we handle it. Right now, I just removed the route handler. But the front end still has a "login-custom" option. Do we remove it from the front end as well, or just display a "functionality coming soon" page?

robertjd replied:
Let's remove this login-custom link from the front-end as well, for now.  We can add this back in when we apply the samples V2 strategy to this repo.

rchild-okta commented on lib/route-handlers.js:
```diff
@@ -25,10 +25,15 @@ const cachedJwks = {};
 
 /**
```
As an aside, we can go ahead and remove the dependencies (like jws, etc) that we no longer need, and also remove them from package.json

vijetmahabaleshwar-okta replied:
Will do.

robertjd replied:
Use eslint to find statements that can be removed from these files.

vijetmahabaleshwar-okta replied:
Ran eslint and removed unwanted declarations and statements.

rchild-okta commented on package.json:
```diff
@@ -38,18 +38,20 @@
     "pem-jwk": "1.5.1",
-    "request": "2.78.0"
+    "request": "2.78.0",
+    "uuid": "^3.1.0"
```
What do we need uuid for?

vijetmahabaleshwar-okta replied:
We don't need it anymore. Will remove it.
As a side-note, is there a way to automatically find the libraries that are no longer needed, by saying running a command?

jmelberg-okta commented on tools/templates/profile.mustache:
```diff
@@ -128,9 +128,11 @@
      {{#user}}
      options.user = {
        email: '{{user.email}}',
-       iss: '{{user.claims.iss}}',
```
Can we keep the `iss` param here? I think it is important to note **who** this is authorized for.

vijetmahabaleshwar-okta replied:
I'm not sure if we still get the "issuer" property in the user object.
We do have the "subject" property which is being displayed but the "issuer" claim is only available in the id_token object which we don't have access to in our backend anymore.

jmelberg-okta replied:
As we discussed yesterday, it might make sense to iterate over the dictionary and pop the `key`, `value` pairs.

See: https://stackoverflow.com/a/17516837/7059401

robertjd replied:
As already mentioned, these templates are going to be over-written by the @okta/samples generator once https://github.com/okta/samples-js-angular-1/pull/24/files is merged

jmelberg-okta commented on tools/mock-okta/util.js:
```diff
@@ -203,10 +204,14 @@ util.mapRequestToCache = (req) => {
-      .replace(`nonce=${query.nonce}`, 'nonce=NONCE');
 
+    // If playing back the tapes, replace the state in req.url with a known state
+    if (!record) {
```
nit: Can we add an `@arg` doc param above to describe what this is?

vijetmahabaleshwar-okta replied:
Done

jmelberg-okta commented on tools/mock-okta/util.js:
```diff
@@ -203,10 +204,14 @@ util.mapRequestToCache = (req) => {
 
+    // If playing back the tapes, replace the state in req.url with a known state
+    if (!record) {
+      req.url = req.url
```
Why are we assigning it to itself?

vijetmahabaleshwar-okta replied:
We're not assigning it to itself. It's actually replacing the state in the URL.
      
req.url = req.url.replace(`state=${query.state}`, `state=${config.state}`);

jmelberg-okta replied:
Hah. My bad. Not sure how I missed that!

jmelberg-okta commented on tools/mock-okta/util.js:
```diff
@@ -224,6 +229,16 @@ util.mapRequestToCache = (req) => {
 
+  // We are redirecting to the client callback. We need to replace the state query parameter with
+  // the state stored in the tape before the redirect
+  else if (req.url.indexOf('/oauth2/v1/authorize/redirect') == 0) {
```
Doesn't this need to include the `/default` authorization server path?

vijetmahabaleshwar-okta replied:
I don't think we need to handle that path as there are no specific parameters passed that would need massaging by the mock server. Is there anything specific you're thinking about in the auth server path, that needs to be handled by the mock server?

jmelberg-okta replied:
Interesting. It does look like Okta removes the *authServerId* for this call. 👍

jmelberg-okta commented:
@vijetmahabaleshwar-okta Have we tested that installing a different frontend, say `@okta/samples-js-react` will still work?

robertjd commented:
It looks like we still have some work to do here:
* If only a subset of the tests are working for now, we need to (1) declare how to run those in this PR and (2) update travis.yml to only run those (so that this PR can pass test requirements)
* Rev the major version of this package, so that the samples generator will not use these new breaking changes
* Run the samples generator to get the new angular templates that were recently merged
* Add a note to the top of this readme that explains: "this sample has been recently updated and will not yet work with all front-end samples, please use the example back-end that is included with any front end sample that you may have cloned"

vijetmahabaleshwar-okta commented:
@robertjd 
* Updated the README with oidc-middleware references
* Removed appserver tests from travis that won't work
* Updated the mustache templates to use the ones from updated angular sample
* Updated the version

jmelberg-okta commented on README.md:
```diff
@@ -97,326 +90,75 @@ Then, replace the *oidc* settings in `.samples.config.json` to point to your new
+### Login Redirect (using [okta oidc-middleware](https://github.com/okta/okta-oidc-js/tree/master/packages/oidc-middleware))
 
-**Note:** Additional configuration for the `SignIn` object is available at [OpenID Connect, OAuth 2.0, and Social Auth with Okta](https://github.com/okta/okta-signin-widget#configuration).
+> Note: The middleware library [okta oidc-middleware](https://github.com/okta/okta-oidc-js/tree/master/packages/oidc-middleware) is used by the Express back-end to handle the [Authorization Code Flow](https://tools.ietf.org/html/rfc6749#section-1.3.1) 
```
nit: We link to this lib in the header too. I'd probably remove the link in the header and keep this.

vijetmahabaleshwar-okta replied:
done

jmelberg-okta commented on README.md:
```diff
@@ -97,326 +90,75 @@ Then, replace the *oidc* settings in `.samples.config.json` to point to your new
+> Note: The middleware library [okta oidc-middleware](https://github.com/okta/okta-oidc-js/tree/master/packages/oidc-middleware) is used by the Express back-end to handle the [Authorization Code Flow](https://tools.ietf.org/html/rfc6749#section-1.3.1) 
 
-### Using a different front-end
+With AngularJS, we include the template directive `ng-click` to begin the login process. When the link is clicked, it calls the `login()` function which simply redirects the client to /login URL. We use the [okta oidc-middleware](https://github.com/okta/okta-oidc-js/tree/master/packages/oidc-middleware) library to handle the login and callback routes at the Express back-end.
```
/login -> `/login`

jmelberg-okta replied:
Can we remove this section about the oidc-middleware?

vijetmahabaleshwar-okta replied:
done

jmelberg-okta commented on README.md:
```diff
@@ -97,326 +90,75 @@ Then, replace the *oidc* settings in `.samples.config.json` to point to your new
+With AngularJS, we include the template directive `ng-click` to begin the login process. When the link is clicked, it calls the `login()` function which simply redirects the client to /login URL. We use the [okta oidc-middleware](https://github.com/okta/okta-oidc-js/tree/master/packages/oidc-middleware) library to handle the login and callback routes at the Express back-end.
 
-By default, this end-to-end sample ships with our [Angular 1 front-end sample](https://github.com/okta/samples-js-angular-1). To run this back-end with a different front-end:
+This middleware library constructs the login redirect URL and performs the redirect on the client's behalf.
```
nit: Not sure we need to note this since it is specific to the front-end.

vijetmahabaleshwar-okta replied:
Do you mean this is under the section "Front-end"?
We need to mention about the URL construction and redirect somewhere. I thought this was the right place to do it.

jmelberg-okta replied:
Isn't it on line `102`? There we explain that the server is performing some "magic" to create and call the `/authorize` endpoint.

vijetmahabaleshwar-okta replied:
You're right. Let me remove this section.

jmelberg-okta commented on README.md:
```diff
@@ -97,326 +90,75 @@ Then, replace the *oidc* settings in `.samples.config.json` to point to your new
+  - Render the user profile information returned by the middleware library
   - Log the user out
 
+> Note: As you can see, the back-end has delegated handling the authorization code exchange, token validation, getting the userinfo etc to [okta oidc-middleware](https://github.com/okta/okta-oidc-js/tree/master/packages/oidc-middleware).
```
This statement is implied by the  section above. We can probably remove this.

vijetmahabaleshwar-okta replied:
I think it's ok to have additional information about this. We don't explicitly mention this anywhere. To a novice developer, this might be helpful.

jmelberg-okta commented on README.md:
```diff
@@ -97,326 +90,75 @@ Then, replace the *oidc* settings in `.samples.config.json` to point to your new
-### Code Exchange
-Next, we exchange the returned authorization code for an `id_token` and/or `access_token`. You can choose the best [token authentication method](http://developer.okta.com/docs/api/resources/oauth2.html#token-request) for your application. In this sample, we use the default token authentication method `client_secret_basic`:
+After successful authentication, we get information about the authenticated user in `req.userinfo` object.
+Below is the code that handles the redirect back after successful authentication
```
authentication -> authentication:

vijetmahabaleshwar-okta replied:
done

viniciusgama commented:
Hey, thanks for the PR, it solves a problem I was having with with the okta mock while using the OIDC Middleware. I even tried to record the HTTP interactions all over again but got stuck in in the auth part. Can someone share how do you do the recording so I can do the same in my own project?

vijetmahabaleshwar-okta commented:
@viniciusgama - You're welcome!
I can share the details on using our mock-server to record your own tapes. But if you ask me, you should be using a real okta org for your needs. Going ahead, we will use the tapes only for our e2e testing. Is there any reason you don't want to use a real okta org?

viniciusgama commented:
hey @vijetmahabaleshwar-okta it's just for local development. And also to make sure that the integration between okta and the middleware are working fine. In our deployed environments we are going to use a real okta instance. You still recommend not having okta for local development?

jmelberg-okta commented on .travis.yml:
```diff
@@ -12,7 +12,7 @@ cache:
     - travis_phantomjs
 before_install:
-  - "npm install -g npm@4.1.2"
+  - "npm install -g npm@5.3.0"
```
Previously we were shrink-wrapping our dependencies, which (I believe) required `npm 4.x`. Since you're using `npm 5`, you might want to commit the `package-lock.json` file.

@robertjd: What do you think?

robertjd replied:
I'm inclined to add this later, after the next round of work (packages are going to change a lot)

jmelberg-okta commented on test/e2e/specs/authorization-code-spec.js:
```diff
@@ -41,7 +41,8 @@ describe('Authorization Code Flows', () => {
 
-  it('can login with a custom login form', () => {
+  // TODO - Enable after hosted login-page flow is integrated 
+  xit('can login with a custom login form', () => {
```
I wonder if we should just remove this? Seems like we will need to revisit this much later.

robertjd replied:
The next step for this repo is to refactor it into the V2 strategy, where we will test this (though the code will probably differ a lot)

vijetmahabaleshwar-okta replied:
Hosted login page for express is already in. So it won't be much long before we integrate it

jmelberg-okta commented on tools/mock-okta/index.js:
```diff
@@ -149,6 +152,45 @@ function handleKeys(req, res) {
+    // Massaging the request headers to match the change in body length
+    req.headers['content-length'] = `${req.body.length}`;
+
+    // Spring back-end doesn't use Basic Auth for token requests whereas express back-end does
```
Curious if this is an implementation on our part, or Spring Security. Both should be exposed.

robertjd replied:
Our token endpoint supports both, but as we're seeing here some clients use the header rather than the body.

vijetmahabaleshwar-okta replied:
AFAIK, it's an implementation in Spring security, not our library.

jmelberg-okta commented on tools/mock-okta/tapes/a15f363b56ad798c9853a981332e6f0e.js:
```diff
@@ -1,7 +1,7 @@
 
 /**
- * GET /oauth2/default/v1/authorize?client_id=0oaom8GeLzZeX7oYZ0g3&redirect_uri=http://localhost:3000/authorization-code/callback&response_type=code&response_mode=query&state=STATE&nonce=NONCE&scope=openid email
+ * GET /oauth2/default/v1/authorize?scope=openid profile email&response_type=code&redirect_uri=http://localhost:8080/authorization-code/callback&state=STATE&client_id=0oap44gsvXUitBZSv0g3
```
Remind me: Does ordering here matter? Do we rebuild this authorize request in mock-okta to ensure the hashes are the same?

robertjd replied:
The authorize request is manually constructed here https://github.com/okta/samples-nodejs-express-4/blob/016cfb650ab743046f15793f32dad57489bcda6f/tools/mock-okta/index.js#L198

vijetmahabaleshwar-okta replied:
Ordering of the query parameters don't matter. But the ordering of the values in the parameters matter.
E.g scope=openid profile will not match a tape recorded with scope=profile openid
